src/blockarray.jl

```diff
@@ -298,10 +298,10 @@ end
 Construct a `BlockMatrix` with `n * m`  blocks.  Each `block_ij` must be an
 `AbstractMatrix`.
 """
-mortar(rows::Vararg{NTuple{M, AbstractMatrix}}) where M =
+mortar(row1::NTuple{M, AbstractMatrix}, rows::Vararg{NTuple{M, AbstractMatrix}}) where M =
     mortar(permutedims(reshape(
-        foldl(append!, rows, init=eltype(eltype(rows))[]),
-        M, length(rows))))
+        foldl(append!, (row1, rows...), init=eltype(row1)[]),
```

Comment:
Do we know `eltype` of `row1` is enough? I would have thought we'd need a `mapreduce(eltype, promote_type, rows)`

Reply:
You're right, I've fixed this now. I've updated this to use the `Base` promotion mechanism instead of trying to guess the `eltype` by ourselves. This often helps with type-inference, so, for example, we now have
```julia
julia> t1 = (ones(Float16, 2,2), ones(Float16, 2,2));

julia> t2 = (ones(Float32, 2,2), ones(Float64, 2,2));

julia> mortar(t1, t2)
2×2-blocked 4×4 BlockMatrix{Float64}:
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
 ──────────┼──────────
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
```
whereas, on master, we obtain
```julia
julia> mortar(t1, t2)
2×2-blocked 4×4 BlockMatrix{Float64, Matrix{AbstractMatrix{Float64}}, Tuple{BlockedUnitRange{Vector{Int64}}, BlockedUnitRange{Vector{Int64}}}}:
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
 ──────────┼──────────
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
```
This has one unintended consequence, though, that sparse matrices may be promoted to dense ones if `Base` decides to change the promotion scheme (as is the case on v1.9 at the moment, although this may be reverted). I'm not sure if this is desired, so I can revert the automatic promotion if we decide against this.
As of now, sparse matrices aren't promoted, unless there is an explicit promotion rule defined by the package, So, for example, the following won't promote the `Fill` to a `Matrix`:
```julia
julia> t1 = (ones(Float16, 2,2), ones(Float16, 2,2));

julia> t2 = (ones(Float32, 2,2), Fill{Float64}(1,2,2));

julia> mortar(t1, t2)
2×2-blocked 4×4 BlockMatrix{Float64, Matrix{AbstractMatrix{Float64}}, Tuple{BlockedUnitRange{Vector{Int64}}, BlockedUnitRange{Vector{Int64}}}}:
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
 ──────────┼──────────
 1.0  1.0  │  1.0  1.0
 1.0  1.0  │  1.0  1.0
```

Reply:
` mortar(t1, t2)` should also convert `t1` and `t2` to the new type.

Reply:
I didn't understand what you're suggesting. Currently it does convert the blocks to a common type, except the common type may be `AbstractMatrix` if there's no promotion rule defined.
```julia
julia> t1 = (ones(Float16, 2,2), ones(Float16, 2,2));

julia> t2 = (ones(Float32, 2,2), ones(Float64, 2,2));

julia> M = mortar(t1, t2);

julia> M.blocks
2×2 Matrix{Matrix{Float64}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]

julia> t2 = (ones(Float32, 2,2), Fill{Float64}(1,2,2))
(Float32[1.0 1.0; 1.0 1.0], Fill(1.0, 2, 2))

julia> M = mortar(t1, t2);

julia> M.blocks
2×2 Matrix{AbstractMatrix{Float64}}:
 [1.0 1.0; 1.0 1.0]  [1.0 1.0; 1.0 1.0]
 [1.0 1.0; 1.0 1.0]  Fill(1.0, 2, 2)
```
Are you saying that promoting the blocks is fine?

Reply:
Creating something of type `BlockMatrix{Float64, Matrix{AbstractMatrix{Float64}}, Tuple{BlockedUnitRange{Vector{Int64}}, BlockedUnitRange{Vector{Int64}}}}` is not ideal as it involves an abstract type so will really hurt performance.

Reply:
I agree, but resolving this would ideally require the matrix types defining their own promotion rules instead of this package deciding on a container for them.

In any case we can discuss this in a separate issue, as this return type has not changed between master and this PR

Reply:
Hmm point taken.

It seems like what we want to support is something more like a tuple-type, but matrix-shaped...there must be some package that has that....